src/Data/List/Properties.agda

```diff
@@ -752,6 +753,36 @@ module _ {P : Pred A p} (P? : Decidable P) where
   ... | true  = cong (x ∷_) (filter-++ xs ys)
   ... | false = filter-++ xs ys
 
+------------------------------------------------------------------------
+-- derun and deduplicate
+
+module _ {R : Rel A p} (R? : B.Decidable R) where
+
+  length-derun : ∀ xs → length (derun R? xs) ≤ length xs
+  length-derun [] = ≤-refl
+  length-derun (x ∷ []) = ≤-refl
+  length-derun (x ∷ y ∷ xs) with does (R? x y) | length-derun (y ∷ xs)
+  ... | true  | r = ≤-step r
+  ... | false | r = s≤s r
+
+  length-deduplicate : ∀ xs → length (deduplicate R? xs) ≤ length xs
+  length-deduplicate [] = z≤n
+  length-deduplicate (x ∷ xs) = ≤-begin
+    1 + length (filter (¬? ∘ R? x) (deduplicate R? xs)) ≤⟨ s≤s (length-filter (¬? ∘ R? x) (deduplicate R? xs)) ⟩
```

Comment:
Here you could save a lot of width by naming `deduplicate R? xs`
in a `let` before the `≤-begin`.